[2단계 - 페이먼츠] 호이초이(최호연) 미션 제출합니다.

hoyyChoi opened the pull request:
동키콩 안녕하세요! 호이초이 입니다! 🤡
엊그제 step1 마무리한 것 같은데, 벌써 step2 리뷰 요청드리네요!
이번에도 깊은 대화 많이 나누면 좋을 것 같습니다!!!!

## 🎯 페이먼츠

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계
- 재사용 가능한 Input Component를 개발한다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 테스트한다.
- 카드 정보를 효과적으로 렌더링 하기 위한 상태 관리를 경험한다.

### 2단계
- 다양한 Form 구성 요소들간의 상태를 효율적으로 관리한다.
- hooks API를 이용하여 상태 관리 로직을 구현하한다.
- custom hooks를 생성하여 Form 관리 로직을 컴포넌트에서 분리하고 재사용한다.
- Controlled & Uncontrolled Components에 입각하여 Form을 핸들링한다.

---

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항(코드 컨벤션, 에러 핸들링 등)을 준수했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 기본 배포 링크 기입: (https://hoyychoi.github.io/react-payments)
  - 스토리북 배포 링크 기입 : (https://6800b7b01be4872cfa12f8a6-zlurvtaypq.chromatic.com/)

### 서비스 구조도 
<img width="500" alt="스크린샷 2025-04-28 오후 2 36 23" src="https://github.com/user-attachments/assets/ebacfde2-1347-4158-b521-cc0708242bbd" />


### 리뷰 요청 & 논의하고 싶은 내용

#### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

1. **컴포넌트 분리** 
컴포넌트의 분리 측면에서 계속 고민을 했던 것 같습니다. 우선, step1을 기반으로 기존의 작업했던 방식을 기반으로 한 번 작업해보기로 하고, 분리를 하지 않았을 때의 불편함을 한 번 느껴보고자 step2에도 이어서 작업을 했습니다. step1에서 제작해놨던 `formUIControllerData`를 기반으로 빠르게 UI를 생성할 수 있었던 것은 큰 장점이었습니다. 하지만 `input`과 `select`처럼 데이터 구조가 다른 컴포넌트를 하나의 흐름으로 다루다 보니 타입 에러가 빈번하게 발생했습니다.
처음에는 `FormSection` 하나로 모두 처리하려 했지만, 타입 충돌 문제를 직접 겪으면서 결국 `FormSection`을 기반으로 `FormSectionInput`과 `FormSectionSelect`를 따로 분리하는 방향으로 리팩토링했습니다. 이 과정을 통해 컴포넌트 분리의 중요성을 다시 한 번 느끼게 되었던 것 같습니다.



2. **포커스 관리**
우선, 포커스의 경우, 
- 1. 현재 필드에서 내부에서 왼쪽 -> 오른쪽으로 이동해야하는 경우
- 2. 필드에서 필드로 이동하는 경우 (아래 -> 위)

이렇게 총 2가지의 포커스 방식에 대해 고려해야한다고 느꼈습니다.
1번째 방식의 경우, `input` 간 포커스를 이동시키기 위해 `ref`를 `InputField` 내부에서 관리했는데, `ref`를 상위 훅에서 관리하는 복잡도와 컴포넌트 내부에서 관리하는 간결성을 비교한 끝에, 현재처럼 `InputField` 내에서 처리하는 것이 적절하다고 판단했습니다.
2번째 방식의 경우, 필드가 바뀔 때 처음 `input`에 자동 포커스를 주는 로직을 추가해 자연스러운 흐름을 유도했습니다. 추가로 포커스 관리에 대한 측면은 아래에 따로 정리해놓았습니다!


#### 2) 이번 리뷰를 통해 논의하고 싶은 부분

1. **컴포넌트의 최적화** 
step1 마지막에 동키콩과 핑퐁을 주고 받으면서, 카드와 관련된 정보인 `useCardInformation` 훅을 각각의 개별 상태로 변경했습니다. 이를 통해 각각의 컴포넌트에서 상태들을 처리하게끔 하고, `React.memo`를 해당 컴포넌트에 씌워 컴포넌트 최적화를 진행하려고 했습니다. 
하지만 `formUIControllerData`를 기반으로 `map`으로 컴포넌트를 렌더링하는 구조 때문에 `step`이 바뀔 때마다 `props`가 달라지고, 결국 리렌더링이 발생했습니다. 이런 경우에 현재 모든 컴포넌트를 모두 호출한 뒤에 이들을 step으로 보여줄지 말지를 조건분기처리를 하고, 각각의 컴포넌트들은 메모이제이션 처리하면 최적화는 될 것 같은데, 가독성 측면에서는 현재 방식이 더 낫다는 생각이 듭니다. 동키콩은 어떻게 생각하시는지 궁금합니다. !

2. **포커스 관리 방법**
`input` 간 이동은 현재 `InputField` 내부에서 `ref` 배열을 만들어 관리하고 있는데, 이 방식이 가장 합리적인지, 아니면 상위에서 `useRef`를 일괄 관리해 `inputField`에 내려주는 패턴이 더 좋은지에 대해 동키콩의 의견을 듣고 싶습니다. 사실 포커스 이동의 경우, 현재 `setState`가 일어나면 해당 value의 길이와 maxLength를 측정하여 포커스를 이동시키고 있지만, 이를 훅에서 함께 관리하면 상태를 바탕으로 포커스도 함께 관리할 수 있으므로, 응집도 측면에서 굉장히 좋을 것이라는 판단이 들었습니다. 또한 각각의 필드 별로 포커스를 관리하면 `autoFocus` 기능을 주지 않더라도 각각 포커스를 관리하여 필드 이동도 자연스러울 것 같다는 생각은 듭니다. 하지만, 이렇게 될 경우 과도한 props drilling이 일어날 것 같기도 한데, 이 부분에 대해서 어떻게 생각하실까요?

3. **`onChange` 로직의 위치**
`cvc`나 `password`의 경우, 단일값이지만, `InputField` 내부의 setState 사용의 통일성을 위해 아래처럼 사용하고자 배열로 관리되고 있습니다. 
```
   setState((prev) => {
      const updated = [...prev];
      updated[index] = value;
      return updated as CardInformationType[T];
    });
```
이를 단일 필드로 변경하면 상위 훅에서 각각의 해당하는 setState 방식을 정의한 후, 각각 내려주는 방식으로 관리하는 게 좋을지, 아니면 `inputField` 레벨에서 그냥 일괄 처리하는 게 나을지 궁금합니다. 

4. **`AddCard`컴포넌트 내부 `handleComplete` 함수 위치**
현재 `AddCard` 페이지 컴포넌트에서 `handleComplete` 함수를 선언하고 `FormContainer`로 내려주고 있는데, 이 함수가 `AddCard`에 있어야 자연스러운지, 아니면 `FormContainer`에 더 가까워야 하는지 궁금합니다!

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도만이 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. Form 상태 관리 & Custom Hook 분리

- 반복되는 로직을 custom hook으로 분리했는가?
- hook 내부와 UI 컴포넌트의 **역할이 명확하게 분리**되어 있는가?
- 상태 흐름이 직관적이며, Form 전체를 일관되게 관리할 수 있는 구조인가?

### 2. 입력 UI 흐름과 UX

- 카드 번호 입력 시 필드 간 **자동 포커싱 이동**이 자연스럽게 동작하는가?
- 숫자만 입력 가능한 필드에서 제한, 에러 메시지, 유효성 피드백 등 **사용자 경험이 충분히 고려**되었는가?

### 3. 컴포넌트 구조 및 재사용성

- 컴포넌트가 명확한 **역할과 책임**을 가지며 과도하게 분리되거나 중첩되어 있지 않은가?
- Input, Button 등 재사용 가능한 컴포넌트가 잘 정의되어 있는가?

### 4. 상태 기반 유효성 검사 및 확인 버튼 활성화

- 모든 필드가 유효할 때만 확인 버튼이 **정확히 활성화/비활성화**되는가?
- 유효성 검사의 기준이 명확하고, **상태 변경에 따른 UI 반응**이 잘 연결되어 있는가?

hoyyChoi committed:
docs: 기능명세서 작성

hoyyChoi committed:
feat: formUIControllerData 기반 map 돌릴때 컴포넌트 key 추가

hoyyChoi committed:
feat: 스토리북 jsDoc 주석 추가 (별도 타입 파일에 추가)

hoyyChoi committed:
feat: select 컴포넌트 제작

hoyyChoi committed:
feat: select 컴포넌트 스토리 제작

hoyyChoi committed:
feat: 비밀번호 입력 필드 추가 및 컴포넌트 생성

hoyyChoi committed:
feat: 비밀번호 입력 컴포넌트 스토리북 제작

hoyyChoi committed:
feat: 비밀번호 value 마스킹 처리

hoyyChoi committed:
feat: 버튼 컴포넌트 제작

hoyyChoi committed:
feat: 버튼 컴포넌트 스토리북 제작

hoyyChoi committed:
feat: completeIcon 이미지 추가

hoyyChoi committed:
docs: 기능명세서 수정

hoyyChoi committed:
style: root 레이아웃 수정

hoyyChoi committed:
feat: select 컴포넌트 App에 추가

hoyyChoi committed:
feat: company 상태 추가 및 information type 추가

hoyyChoi committed:
feat: 카드사 색상 정리

hoyyChoi committed:
feat: 카드사에 따른 색상 변경 로직 추가(PreviewCard)

hoyyChoi committed:
style: select 폰트 align 수정

hoyyChoi committed:
feat: 동적 ui 생성 테스트

hoyyChoi committed:
feat: step별 동적ui 구현을 위한 에러 상태 판별

hoyyChoi committed:
refactor: 카드 정보 상태 분리

hoyyChoi committed:
feat: 카드 정보 별 상태 분배 및 FormSection 분리 및 타입 재정의

hoyyChoi committed:
fix: useCheckComplete 함수 네이밍 변경 및 파일 위치 이동

hoyyChoi committed:
feat: maxLength 상수 처리

hoyyChoi committed:
feat: useCheckLengthComplete 훅 생성

hoyyChoi committed:
feat: useCheckLengthComplete 훅 이용 상태 별 complete 확인하기

hoyyChoi committed:
feat: useStep 훅으로 동적 ui 구현

hoyyChoi committed:
feat: 모든 필드 입력 + 유효성 통과시, 버튼 활성화 및 useAllComplete 훅 생성

hoyyChoi committed:
style: 레이아웃 css 수정

hoyyChoi committed:
chore: react-router 의존성 설치

hoyyChoi committed:
chore: 폴더명 변경

hoyyChoi committed:
style: 전체 스타일 파일 위치 변경

hoyyChoi committed:
feat: react-router를 이용하여 페이지 관리

hoyyChoi committed:
chore: App파일 삭제

hoyyChoi committed:
style: Index.css 위치 변경

hoyyChoi committed:
chore: 폰트 직접 호스팅

hoyyChoi committed:
feat: completeCard 컴포넌트 제작

hoyyChoi committed:
chore: favicon 추가

hoyyChoi committed:
feat: 확인 버튼 누를시, 페이지 이동 및 데이터 전달

hoyyChoi committed:
docs: 기능명세서 수정

hoyyChoi committed:
refactor: 폼 제출 -> 확인 버튼 FormContainer 내부로 이동

hoyyChoi committed:
feat: 사용자 입력시 자동 포커싱 기능 구현

hoyyChoi committed:
fix: storybook 에러 수정

hoyyChoi committed:
chore: 공통 훅 common폴더 분리

hoyyChoi committed:
refactor: 컴포넌트 props 네이밍 변경 및 통일 (setter-> setState로 변경)

hoyyChoi committed:
refactor: 타입 정리

hoyyChoi committed:
fix: storybook 에러 해결

hoyyChoi committed:
fix:  createBrowserRouter -> BrowserRouter로 변경

JUDONGHYEOK commented on src/types/componentPropsType.ts:
```diff
@@ -1,29 +1,62 @@
-import { useEachValidationType, useValidationType } from "./useValidationType";
+import { useEachValidationType, validationFieldType } from "./useValidationType";
 
 export type FormContainerProps = {
```
여기에 있는 type들은 component에 바로 위치해도 괜찮지 않나 생각이 되는데요. 그 이유는 사용처가 컴포넌트밖에 없기도 하고 story파일에서 사용되더라도 컴포넌트의 prop type이기 때문에 컴포넌트 내부에 위치하는게 모듈의 응집도가 높고 자연스럽다고 생각해서요

hoyyChoi replied:
컴포넌트 props의 타입을 컴포넌트가 가져가는게 응집도 측면과 더불어 수정 및 확장에 용이할 것 같습니다! 수정해보겠습니다!

hoyyChoi replied:
동키콩! 수정하는 과정에서 궁금한 점이 생겨 여쭤봅니다!
컴포넌트에 해당하는 `props` 타입을 컴포넌트 파일 내부로 옮기려다 보니, 예를 들어 `FormSectionInput`의 `props` 중 `fieldData: InputFieldProps<T>`처럼 다른 컴포넌트에서 선언된 타입을 참조하는 경우, 다시 import를 해줘야 합니다!
이런 경우에도 `props` 타입은 각 컴포넌트 내부에 선언해두는 게 맞을까요? 혹은 공통 타입은 별도 파일로 관리하는 것이 더 나을지 고민이 됩니다!

JUDONGHYEOK replied:
기본적으로 컴포넌트타입들은 컴포넌트내부에서 정의하되 범용적으로 사용되는 타입이 있다면 추출해서 commonType에 위치시켜도 좋겠네요!

JUDONGHYEOK commented on src/components/Select/Select.tsx:
```diff
@@ -0,0 +1,117 @@
+  // 모달 열고 닫기 상태
+  const [isOpen, setIsOpen] = useState(false);
+
+  // 카드 선택이 되었을 때 상태
+  const [isSelect, setIsSelect] = useState(false);
+
+  // select의 라벨에 보여줄 데이터
+  const [selected, setSelected] = useState(placeholder);
```
여기는 모두 동일하게 select가 선택되었는지 선택되지 않았는지를 나타내는 값이라 하나의 상태에서 추출가능한 값이라 생각이 되는데요. 
하나의 상태로 줄일 수도 있을 것 같고 아니면 Select에서 value라는 prop을 뚫어줘서 판단할 수도 있을 것 같아요. 현재는 상태가 두개를 둘 필요는 없을 것 같습니다!

hoyyChoi replied:
지금 보니, isSelect는 불필요한 상태였군요,,! 최대한 신경쓴다고 했는데, 집중을 못했던 것 같습니다! selected와 placeholder 값을 비교해서 isSelect 값을 사용해보도록 하겠습니다!

hoyyChoi replied:
[628cf8a](https://github.com/woowacourse/react-payments/pull/481/commits/628cf8a61a7d16a46e7102f6b8f25c5f957efb14)

JUDONGHYEOK commented on src/components/FormContainer/FormContainer.tsx:
```diff
@@ -1,35 +1,82 @@
+        .slice(0, step + 1)
+        .reverse()
+        .map((formSectionData) => {
+          if (formSectionData.key === "company") {
```
의미상으로는 type값을 확인하는 것이 맞지 않나? 생각이 되네요.
또 사실 custom한 select를 사용하기 때문이기도 하지만 input type이 하나만 추가되는 것이 이렇게 로직이 복잡해져서 조금 고민이 필요하겠네요. 제 생각에는 input(input,select)같은 아이들의 인터페이스를 어느정도 통일 시켜두고 input type을 판단하는 중간 컴포넌트를 만들어 보는 것도 좋지 않을까 생각이 되어요.
말로 하니까 조금 횡설수설인데 코드로 보면 약간 아래 스타일 처럼요.

```typescript
// inputType을 따로 정의
export const formType: Record<InputType, React.ReactElement> = {
    input: FormSectionInput,
    select: FormSectionSelect
}


const FieldRenderer = ({ type, ...rest }: { type: InputType }) => {

    const FieldComponent = formType[type];

    return (
        <FieldComponent {...rest} />
    );
};
```

hoyyChoi replied:
동키콩!
우선, 말씀해주신 내용을 반영하여 `FieldRenderer` 함수를 만들고 이를 통해 컴포넌트를 분기하는 방식으로 리팩토링을 진행해보았습니다.

리뷰에서 언급하신 `input(input, select) 같은 아이들의 인터페이스를 어느정도 통일시켜두고`라는 부분은, 혹시 내부에서 받는 `props` 구조를 통일시켜라라는 의미일까요?!
만약 그렇다면, 현재 `input`과 `select`는 서로 받는 상태 관련 데이터가 다르기 때문에 `props` 구조를 완전히 동일하게 맞추는 것은 현실적으로 어려울 것 같다는 판단이 들었습니다. 오히려 이들을 하나의 컴포넌트로 통합하거나, 타입 분기 처리를 통해 공통화하려고 하다 보니 오히려 복잡도가 올라가고 단점이 더 두드러지는 것 같다는 느낌도 있습니다.

실제로 리팩토링 과정에서 `React.FC`의 제네릭 타입 지정에서도 타입 에러가 계속 발생해 꽤 시간을 들여 고민했지만, 결국 현재는 일단 any로 처리해둔 상태입니다. (타입스크립트에 대한 이해가 많이 부족한 것 같습니다,,)

또한 `FieldRenderer`는 로직이 단순하고, 실질적으로 `FormContainer`에 종속된 컴포넌트라고 판단되어 현재는 내부에 함께 위치시켜두었습니다.! 특정 컴포넌트에서만 사용되는 단순 분기 컴포넌트의 경우, 파일 분리보다는 해당 컴포넌트 내부에 함께 관리하는 것이 더 자연스러운지 궁금합니다!

hoyyChoi replied:
[8a97e1c](https://github.com/woowacourse/react-payments/pull/481/commits/8a97e1c83e121565d23613e20019d52793eee4c8)

JUDONGHYEOK replied:
인터페이스를 통합하여 뚫어주는 것이 복잡도가 올라간다고 판단하셨다면 prop에 대한 분기도 FieldRenderer에 포함시키는 것은 어떻게 생각하세요? map내부 로직이 복잡해서 오히려 역할을 나눠주는 것이 좋을 수도 있을 것 같아서요!

hoyyChoi replied:
말씀하신대로 필드별 `props` 조립 로직까지 `FieldRenderer` 컴포넌트에 위임하도록 리팩토링했습니다. 
이에 따라 `FormContainer`의 가독성이 좋아지고 역할 분리가 명확해졌습니다.
기존에는 단순히 `type`에 따라 `input` 또는 `select` 컴포넌트를 분기하는 역할만 했던 구조였지만, 분기처리를 위임함으로써, `FieldRenderer`의 책임이 조금 무거워졌기 때문에, 별도 파일로 분리하여 관리하도록 구성했습니다.!

[dd08c17](https://github.com/woowacourse/react-payments/pull/481/commits/dd08c1780a8131a69f19cb4279c30cecce52c910)

JUDONGHYEOK commented on src/hooks/common/useCardType.tsx:
중요한 것은 아니지만 jsx문법을 사용하지 않고 있기 때문에 hook들의 확장자 명은 ts로 해도 괜찮을 것 같습니다.

hoyyChoi replied:
[f4c2b24](https://github.com/woowacourse/react-payments/pull/481/commits/f4c2b24d30ba8499629f8afdbb2983c0bebcc248)

JUDONGHYEOK commented on src/hooks/common/useCheckLengthComplete.tsx:
```diff
@@ -0,0 +1,9 @@
+import { useMemo } from "react";
+
+export const useCheckLengthComplete = <T extends { length: number }>(state: T[], maxLength: number): boolean => {
```
여기 제네릭을 뚫지 않고 그냥 state에 타입을 직접 넣어줘도 괜찮지 않을까요?? 타입을 직접 넣어주는 것 만으로도 충분히 의미가 전달되기도 하구요

hoyyChoi replied:
저는 처음엔 제네릭으로 처리하면 타입이 더 유연하고 정확할 거라고 생각해서 `<T extends { length: number }>` 형태로 작성했었습니다.
그런데 지금 생각해보니까, 이미 제가 정의해둔 `UniqueNumberType`, `ExpirationDateType`, `CvcNumberType` 등의 타입들이 있기 때문에, 이걸 유니언 타입으로 묶어서 처리하면 더 명확하고 의도에 잘 맞는 타입 제한이 될 것 같습니다.

hoyyChoi replied:
[692ef36](https://github.com/woowacourse/react-payments/pull/481/commits/692ef3626a48c0e554610ea19f9f545af42741a8)

JUDONGHYEOK commented on src/components/Input/Input.tsx:
```diff
@@ -1,18 +1,24 @@
-    />
-  );
-};
+const Input = forwardRef<HTMLInputElement, InputProps>(
```
React 19를 사용하고 계시니 forwardRef는 필요하지 않을 것 같아요

hoyyChoi replied:
아하 그렇군요! 좋은 정보 감사합니다!!!

hoyyChoi replied:
[cf4ed74](https://github.com/woowacourse/react-payments/pull/481/commits/cf4ed74abbf67fe29384c6ef2bac40e02d8f1cea)

JUDONGHYEOK commented on src/hooks/common/useAllComplete.tsx:
```diff
@@ -0,0 +1,15 @@
+
+const useAllComplete = (isStateCompletes: isStateCompletesType, isErrorCompletes: isErrorCompletesType) => {
+  return useMemo(() => {
+    const isAllStateCompletes = Object.values(isStateCompletes).every((bool) => bool === true);
```
every내의 callback 변수라도 bool은 너무 추상적이라 조금더 의미를 명확하게 하면 좋을 것 같아요
```typescript
    const isAllStateCompletes = Object.values(isStateCompletes).every((isComplete) => isComplete);
```

hoyyChoi replied:
[fd073a1](https://github.com/woowacourse/react-payments/pull/481/commits/fd073a14947e7de2b13fa695609c257d77e8c0c7)

JUDONGHYEOK commented on src/hooks/common/useStep.tsx:
```diff
@@ -0,0 +1,27 @@
+const useStep = (isStateCompletes: isStateCompletesType, isErrorCompletes: isErrorCompletesType) => {
+  const [step, setStep] = useState(0);
+
+  useEffect(() => {
+    const currentKey = STEP_ORDER[step];
+    const isStateComplete = isStateCompletes[currentKey];
+    const isErrorComplete = currentKey === "company" ? true : isErrorCompletes[currentKey];
+
+    if (isStateComplete && isErrorComplete) {
+      setStep((prev) => {
+        if (prev < STEP_ORDER.length - 1) {
+          return prev + 1;
+        }
+        return prev;
+      });
+    }
+  }, [isStateCompletes, isErrorCompletes]);
```
여기 함수의 depth가 깊어지고 하는일이 조금 많아진 것 같아 아래와 같이 분리해볼 수 있을 것 같아요
```typescript
  const checkCurrentStepComplete = useCallback((currentStep: number) => {
    if (currentStep >= STEP_ORDER.length) return false;

    const currentKey = STEP_ORDER[currentStep];
    const isStateComplete = isStateCompletes[currentKey];

    const isErrorComplete =
      currentKey === "company" || (isErrorCompletes[currentKey] ?? false);

    return isStateComplete && isErrorComplete;
  }, [isStateCompletes, isErrorCompletes]);

  useEffect(() => {
    if (checkCurrentStepComplete(step) && step < STEP_ORDER.length - 1) {
      setStep(step + 1);
    }
  }, [step, checkCurrentStepComplete]);
```

hoyyChoi replied:
동키콩! 이렇게 하면 조금 더 깔끔한 코드를 만들 수 있을 것 같습니다! `useCallback`으로 함수 자체를 메모이제이션 하는 것도 좋은 방법인 것 같습니다! 
다만 궁금한 점이 하나있습니다! 지금 현재 예시로 주신 코드를 보면 `useEffect`의 의존성 배열에 step이 들어가 있는데, 이 값은 굳이 넣지 않아도 되는거라고 판단되는데, 맞을까요?

`checkCurrentStepComplete` 자체에서 `isStateCompletes`, `isErrorCompletes` 상태 값이 변경되면 해당 함수가 다시 재정의 되기 때문에 이 함수만 `useEffect`의 의존성 배열로 정의해놓으면 될 것 같은데 맞을까요?

hoyyChoi replied:
그리고 현재 `checkCurrentStepComplete`함수의 인자값으로  `currentStep`을 받고 있는데, 이건 상위 state에서 참조할 수 있는 값이라고 판단이 듭니다! 그럼에도 인자로 넘겨줌으로써 얻을 수 있는 장점있을까요?
단순히, 외부 상태인 `step`을 참조할 필요 없이 인자로 넘기는게 명확하고 깔끔해서인지 궁금합니다!

hoyyChoi replied:
[52387ff](https://github.com/woowacourse/react-payments/pull/481/commits/52387ffea186c81d7f45d9d2421b5b14c4a77d43)

JUDONGHYEOK replied:
오 그렇네요! checkCurrentStepComplete함수를 별도의 모듈로 분리한다던가 그렇게 된다면 의미가 있겠지만 지금은 step을 직접참조해도 무방할 것 같습니다!

hoyyChoi replied:
[3b5522e](https://github.com/woowacourse/react-payments/pull/481/commits/3b5522e14290aff48463a4b88af56e74ccf9ddcb)

JUDONGHYEOK commented on src/components/FormContainer/FormContainer.tsx:
```diff
@@ -1,35 +1,82 @@
-      })}
-    </div>
+    <form css={FormContainerStyle} onSubmit={onSubmit}>
+      {formUIControllerData
+        .slice(0, step + 1)
+        .reverse()
+        .map((formSectionData) => {
```
조금 개선을 해본다면 현재는 모든 prop이 변할때 해당 로직들이 실행되고 있습니다. useMemo와 같은 메모이제이션을 활용한다면 성능적으로도 종속성이 변경될 때만 재계산되기 때문에 이점이 있을 것이고 가독성 측면에서도 의존성을 명시해주기 때문에 이점을 얻을 수 있을 것이라 생각합니다.

hoyyChoi replied:
현재 `formUIControllerData` 재반복하면서 `fieldData`가 새로운 `props`로 들어가 해당 컴포넌트 내에서는 새로운 값이라고 판단을 하여 `memoization`이 안되는 것으로 알고 있는데, 이게 맞을까요?

동키콩이 말씀하신 부분이 `useMemo`를 `fieldData`에 적용하여 메모이제이션을 하라는 말씀이실까요? `FormSectionInput` 에 콘솔을 출력하여 디버깅을 해보았지만, 메모이제이션이 안되는 것 같습니다..

어떻게 하면 현재 부분에서 메모이제이션을 가능하게 할 수 있는지 구체적인 설명 부탁드려도 괜찮을까요!? (제가 이해를 잘 못했습니다,!)

JUDONGHYEOK replied:
아 아래와 같이 step이 바뀔때만 해당 로직이 실행되도록 메모이제이션을 하는 것이 어떠신지 제안드린 것이였어요!

```typescript
const visibleFields = useMemo(() => {
  return formUIControllerData
    .slice(0, step + 1)
    .reverse();
}, [step]);
```

hoyyChoi replied:
너무 좋은 것 같습니다! 생각해보니까, step이 안바뀌더라도 상위 상태들이 바뀌면 현재 계속 slice와 reverse 작업을 진행하고 있었군요! 
수정완료 했습니다. [7fbb28c](https://github.com/woowacourse/react-payments/pull/481/commits/7fbb28cc9b902e4922ea519a1bbf83244d06e27a)

JUDONGHYEOK commented on src/components/InputField/InputField.tsx:
```diff
@@ -32,11 +37,16 @@ const InputField = ({
             maxLength={inputProps.maxLength}
             error={isError[index]}
+            masking={inputProps.masking}
+            autoFocus={index === 0}
```
오 autofocus의 활용이 좋네요!👍

JUDONGHYEOK commented on src/components/FormSection/FormSectionSelect.tsx:
```diff
@@ -0,0 +1,20 @@
+};
+
+export default React.memo(FormSectionSelect) as typeof FormSectionSelect;
+// export default FormSectionSelect;
```
주석은 지워주세요!

hoyyChoi replied:
[432be1d](https://github.com/woowacourse/react-payments/pull/481/commits/432be1ddd850d23562c35835d331f9d1db351543)

JUDONGHYEOK reviewed (changes requested):
안녕하세요 호이초이!!! 2단계도 깔끔하게 기능들을 구현해주셨군요. 수고 많으셨습니다. 1단계에서 마지막에 드린 리뷰들도 다 잘 반영해주셔서 감사합니다! 컴포넌트들의 역할들이 분리되고 깔끔해졌다는 느낌을 받았어요. 도메인들도 custom hook들로 분리되어 구조를 파악하기 용이했던 것 같구요.


 새로 추가된 기능들에 대해서 기존 구조가 어떤 문제점을 가지고 있었는지에 관해서 고민하시는 부분이 굉장히 인상적이네요. 지금 구조의 문제점을 파악하는데 가장 좋은 방법은 잘 변하는 부분과 잘 변하지 않는 부분을 파악하는 것부터 시작이라고 생각하는데요. OCP를 기반으로 기능을 추가함에 있어서 기존 코드를 얼마나 건드려야 하는지를 고민해보시면 좋을 것 같아요!

## 2) 이번 리뷰를 통해 논의하고 싶은 부분
> 컴포넌트의 최적화
step1 마지막에 동키콩과 핑퐁을 주고 받으면서, 카드와 관련된 정보인 useCardInformation 훅을 각각의 개별 상태로 변경했습니다. 이를 통해 각각의 컴포넌트에서 상태들을 처리하게끔 하고, React.memo를 해당 컴포넌트에 씌워 컴포넌트 최적화를 진행하려고 했습니다.
하지만 formUIControllerData를 기반으로 map으로 컴포넌트를 렌더링하는 구조 때문에 step이 바뀔 때마다 props가 달라지고, 결국 리렌더링이 발생했습니다. 이런 경우에 현재 모든 컴포넌트를 모두 호출한 뒤에 이들을 step으로 보여줄지 말지를 조건분기처리를 하고, 각각의 컴포넌트들은 메모이제이션 처리하면 최적화는 될 것 같은데, 가독성 측면에서는 현재 방식이 더 낫다는 생각이 듭니다. 동키콩은 어떻게 생각하시는지 궁금합니다. !

저도 지금 수준이 적당하다고 생각합니다. 과도한 메모이제이션은 오히려 가독성을 떨어뜨릴 수 있다고 생각해요. 현재 구조에서는 step 변경에 따른 리렌더링이 불가피한 면이 있으며, 이 정도 규모의 프로젝트에서는 성능 측면에서도 큰 문제가 되지 않을 거라고 생각해요. 조금 더 개선할 수 있다고 생각하는 부분은 코멘트에 남겨두었으니 확인 부탁드립니다!

> 포커스 관리 방법
input 간 이동은 현재 InputField 내부에서 ref 배열을 만들어 관리하고 있는데, 이 방식이 가장 합리적인지, 아니면 상위에서 useRef를 일괄 관리해 inputField에 내려주는 패턴이 더 좋은지에 대해 동키콩의 의견을 듣고 싶습니다. 사실 포커스 이동의 경우, 현재 setState가 일어나면 해당 value의 길이와 maxLength를 측정하여 포커스를 이동시키고 있지만, 이를 훅에서 함께 관리하면 상태를 바탕으로 포커스도 함께 관리할 수 있으므로, 응집도 측면에서 굉장히 좋을 것이라는 판단이 들었습니다. 또한 각각의 필드 별로 포커스를 관리하면 autoFocus 기능을 주지 않더라도 각각 포커스를 관리하여 필드 이동도 자연스러울 것 같다는 생각은 듭니다. 하지만, 이렇게 될 경우 과도한 props drilling이 일어날 것 같기도 한데, 이 부분에 대해서 어떻게 생각하실까요?

focus를 관리하는 위치는 지금처럼 InputField에서 관리하는 것이 좋다고 생각해요. 포커스 관리는 도메인 로직이랑 섞일 필요가 없는 UI 로직이라고 생각이 되어서요.
 이와 별개로 현재처럼 util을 활용하는 방법과 custom hook을 통해 관리하는 방법이 있을 것 같은데요. moveFocusOrBlur도 결국엔 ref를 전달받아야 하므로 포커스 관리를 위한 전용 커스텀 훅을 만들어 재사용성을 높이는 것이 좋지 않을까 생각이 됩니다.

> onChange 로직의 위치
cvc나 password의 경우, 단일값이지만, InputField 내부의 setState 사용의 통일성을 위해 아래처럼 사용하고자 배열로 관리되고 있습니다.
이를 단일 필드로 변경하면 상위 훅에서 각각의 해당하는 setState 방식을 정의한 후, 각각 내려주는 방식으로 관리하는 게 좋을지, 아니면 inputField 레벨에서 그냥 일괄 처리하는 게 나을지 궁금합니다.

저도 cvc나 password같은 것은 단일 필드로 관리되는 것이 더 자연스럽다고 봅니다. 따라서 InputField에서는 공통의 인터페이스만 관리하고 실제 업데이트 로직은 상위에서 제공받는 것이 더 적절할 것 같아요.

그런데 지금 구조에서 단일 필드로 변경하면 변경점이 조금 많을 것 같아 우려가 되기는 하네요. 이러한 변경은 데이터 모델과 타입 정의부터 시작해서 여러 파일에 영향을 줄 수 있을 것 같은데요. 한번 시도해보시는 것은 좋을 것 같아요!

> AddCard컴포넌트 내부 handleComplete 함수 위치
현재 AddCard 페이지 컴포넌트에서 handleComplete 함수를 선언하고 FormContainer로 내려주고 있는데, 이 함수가 AddCard에 있어야 자연스러운지, 아니면 FormContainer에 더 가까워야 하는지 궁금합니다!

각각의 장단점이 있을 것 같아요. AddCard에 둔다면 cardInformationState에 직접 접근할 수 있어 데이터 전달이 간단하기도 하고 FormContainer가 라우팅 로직에 의존하지 않으므로 재사용성이 올라가지만 Prop이 추가되는 문제가 있죠. 또 FormContainer에 위치하면 props driling이 감소하지만 재사용성이 저하되고 FormContainer의 관심사가 혼합될 수 있기도 하구요.

큰 차이는 없을 것 같지만 제 개인적인 의견으로는 FormContainer보다는 AddCard에 두는 것이 컴포넌트 이름에서 오는 역할로는 더 적합한 것 같아요.

---
2단계도 너무 수고많으셨어요!! 이미 좋은 코드라고 생각되어서 리뷰가 많지는 않네요.🥲 확인해주시고 다시 요청주시면 감사하겠습니다!!

hoyyChoi committed:
docs: 기능명세서 수정

hoyyChoi committed:
fix: Input 컴포넌트 forwardRef 제거

hoyyChoi committed:
refactor: useCheckLengthComplete 인자 state 타입 명시적으로 변경

hoyyChoi committed:
chore: hook 확장자 명 ts 수정

hoyyChoi committed:
fix: Select 컴포넌트 불필요한 isSelect 상태 제거

hoyyChoi committed:
fix: useAllComplete 훅 every 메소드 bool 네이밍 명시적으로 변경

hoyyChoi committed:
chore: FormSectionSelect 컴포넌트 export 구문 주석제거

hoyyChoi committed:
refactor: useStep 훅의 useEffect 내부 로직 useCallback으로 분리

hoyyChoi committed:
refactor: useAllComplete 훅 every 메소드 bool 네이밍 명시적으로 변경

hoyyChoi committed:
feat: Inputfield 컴포넌트 moveFocusOrBlur 훅으로 분리

hoyyChoi committed:
refactor:  FieldRenderer 컴포넌트 제작 후, FormContainer 로직 제어

hoyyChoi committed:
refactor:  cardInformation 단일 값 분리 (cvc, password)

hoyyChoi committed:
fix: useCheckLengthComplete 훅 상태 배열 및 단일값 분기처리

hoyyChoi committed:
refactor: InputField에서 공통 인터페이스 관리 및 실제 업데이트 로직 상위에서 제공받기

hoyyChoi committed:
refactor: 훅 결합 (상태 훅 + 에러 훅 결합)

hoyyChoi committed:
refactor: 에러 상태 단일값 cvc, password -> 변경

hoyyChoi committed:
fix: storybook 에러 해결

hoyyChoi commented:
안녕하세요, 동키콩🤡
페이먼츠 모듈 미션과 일정이 겹치면서 수정 기간이 예상보다 길어졌습니다. 죄송합니다 🥲
모듈 미션을 진행하면서 현재 수정 중인 코드에 대해서도 많은 고민을 하게 되었고, 그 과정에서 대대적인 수정을 거치게 되어 코드가 많이 바뀌었습니다.!
이번에도 리뷰 잘 부탁드리겠습니다!

## 피드백 답변
>이와 별개로 현재처럼 util을 활용하는 방법과 custom hook을 통해 관리하는 방법이 있을 것 같은데요. moveFocusOrBlur도 결국엔 ref를 전달받아야 하므로 포커스 관리를 위한 전용 커스텀 훅을 만들어 재사용성을 높이는 것이 좋지 않을까 생각이 됩니다.

저도 같은 생각입니다! 내부에서 `useRef`를 사용한다면 커스텀 훅으로 분리해 재사용할 수 있기 때문에, 포커스 관리를 위한 훅으로 구성하는 것이 더 좋은 방향이라고 생각합니다.
**수정 커밋** : [e01fbff](https://github.com/woowacourse/react-payments/pull/481/commits/e01fbffa7a86989a5778014d586feb0200e66930)


> 저도 cvc나 password같은 것은 단일 필드로 관리되는 것이 더 자연스럽다고 봅니다. 따라서 InputField에서는 공통의 인터페이스만 관리하고 실제 업데이트 로직은 상위에서 제공받는 것이 더 적절할 것 같아요.
그런데 지금 구조에서 단일 필드로 변경하면 변경점이 조금 많을 것 같아 우려가 되기는 하네요. 이러한 변경은 데이터 모델과 타입 정의부터 시작해서 여러 파일에 영향을 줄 수 있을 것 같은데요. 한번 시도해보시는 것은 좋을 것 같아요!

사실 이 부분을 수정하느라 꽤 큰 수정이 있었습니다. 단순히 초기값만 바꾸는 수준이 아니라, 연관된 `useCheckLengthComplete`, `useErrorCheckComplete` 훅들과 타입 정의까지 함께 수정해야 했고, 이를 기반으로 화면에 보여주는 값들도 분기 처리를 통해 다시 구성해야 했습니다.
또한 상태값뿐만 아니라 에러 상태 역시 배열에서 단일 boolean 값으로 변경하는 작업도 병행했습니다. 결국 처음부터 급하게 설계한 제 책임이라고 생각하며, 이번 수정을 통해 초기 설계의 탄탄함을 조금 느낀 것 같습니다..!
**수정 커밋** : 상태 분리 ([ed7fa2e](https://github.com/woowacourse/react-payments/pull/481/commits/ed7fa2e3d3354745c0c3b643914f0d96dc14134b), [2ef86ec](https://github.com/woowacourse/react-payments/pull/481/commits/2ef86eca85d7df689aa4f5b66581fef180e7b5a1), [4432539](https://github.com/woowacourse/react-payments/pull/481/commits/443253962231bcd91f691ef39a861d40bd619bea)) / 에러 분리 ([c01ea5a](https://github.com/woowacourse/react-payments/pull/481/commits/c01ea5a6669d7bc2f072e2cbdfc85b0bb935ad9d))


## 추가 수정 사항
### **훅 결합 (상태 훅 + 에러 훅)**
1단계 마지막 수정 당시에는 상태 훅과 에러 훅을 분리한 뒤, 이를 컴포넌트에서 결합해 사용하는 방향으로 설계를 진행했습니다. 당시에는 괜찮은 방식이라 판단했지만, 이번 페이먼츠 모듈 미션을 진행하며 그 설계가 조금 잘못되었다는 것을 느꼈습니다.

에러 훅은 상태 훅과 밀접한 연관이 있으며, 상태에 따라 에러가 달라지기 때문에 이를 분리해 관리하는 것은 응집도 측면에서 적절하지 않다는 판단이 들었습니다. 실제로 페이먼츠 모듈 미션에서는 에러 훅만 따로 만들어 모듈로 배포했지만, 페이먼츠 커스텀 훅의 본질은 단순한 유효성 검사 함수가 아니라 **입력 필드의 상태 관리와 유효성 검사 로직을 함께 포함하는 '입력 단위 훅'**이라고 생각합니다.

훅은 입력에 필요한 상태와 로직을 모두 포함하고 있어야 하며, UI 컴포넌트에서는 해당 훅을 연결하는 것만으로도 최소한의 코드로 동작하는 입력 필드를 만들 수 있어야한다는 생각이 들었습니다!

이러한 고민 끝에, 이번에는 상태와 에러 메시지, 에러 상태를 모두 포함하고, setter와 함께 에러가 모두 통과했는지 여부와 입력값의 길이가 유효한지를 반환하는 하나의 커스텀 훅으로 통합하여 설계해 보았습니다.
이 방식에 대해서 동키콩은 어떻게 생각하시는지 궁금합니다!!

**수정 커밋** : [9256339](https://github.com/woowacourse/react-payments/pull/481/commits/925633905fddb1b3e3071fdc87a04843e928ef31)

### 고민 사항
앞서 상태를 단일값과 배열값으로 변경하면서, 각각의 `setter` 함수를 `setState` 커스텀 함수 형태로 커스텀 훅 내부에서 정의하고, 이를 상위 컴포넌트로 전달하는 방식으로 리팩토링을 진행했습니다.
또한 에러 훅과 상태 훅을 결합하면서, 자연스럽게 유효성 검증과 에러 상태 및 메시지를 함께 처리하는 `validateInput` 함수도 해당 훅 내부에서 관리하도록 구성했습니다.

기존에는 상태훅과 에러훅이 분리되어 있었기 때문에 `setter`와 `validateInput`을 `inputField` 컴포넌트에서 직접 조합해 `onChange`에 전달해야 했습니다. 하지만 현재는 동일한 훅 내부에서 이 두 로직을 통합해 내려주고 있습니다.

이와 같은 흐름이라면, 처음부터 훅 내부에서 setter + 유효성 검증 로직을 조합해 하나의 `onChange` 핸들러로 내려주는 것이 더 깔끔한 설계일 것 같다는 생각이 듭니다. 혹시 이에 대해 동키콩은 어떻게 생각하실까요!?

만약 이 방식이 더 적절하다면, 해당 핸들러 함수의 네이밍을 어떻게 정해야 할지도 고민이 됩니다. 예를 들어 단순한 `handleChange`보다 역할이 명확히 드러나는 이름으로 짓는 것이 좋을 것 같다고 생각 중입니다.

```js
 /** setter(상태업데이트) */
  const setUniqueNumber: SetValueFn<UniqueNumberType[number]> = (value, index) => {
    if (index === undefined) return;

    /** 에러 및 에러메세지 매칭 */
    // validateInput(value, index);

    // setter 상태업데이트
    setUniqueNumberState((prev) => {
      const updated = [...prev];
      updated[index] = value;
      return updated as UniqueNumberType;
    });
  };

  const validateInput = (value: string, index: number) => {
    const { error, message } = validateNumberOnly(value);

    setIsError((prev) => {
      const updated = [...prev];
      updated[index] = error;
      return updated;
    });

    setErrorMessage(message);
  };

...

return {
    setState: setUniqueNumber,
    validateInput,
    ...
  };
```

JUDONGHYEOK commented:
질문에 대한 제생각을 여기 코멘트로 남겨보겠습니다!

> > 훅 결합 (상태 훅 + 에러 훅)
> 1단계 마지막 수정 당시에는 상태 훅과 에러 훅을 분리한 뒤, 이를 컴포넌트에서 결합해 사용하는 방향으로 설계를 진행했습니다. 당시에는 괜찮은 방식이라 판단했지만, 이번 페이먼츠 모듈 미션을 진행하며 그 설계가 조금 잘못되었다는 것을 느꼈습니다.
> 
> 에러 훅은 상태 훅과 밀접한 연관이 있으며, 상태에 따라 에러가 달라지기 때문에 이를 분리해 관리하는 것은 응집도 측면에서 적절하지 않다는 판단이 들었습니다. 실제로 페이먼츠 모듈 미션에서는 에러 훅만 따로 만들어 모듈로 배포했지만, 페이먼츠 커스텀 훅의 본질은 단순한 유효성 검사 함수가 아니라 **입력 필드의 상태 관리와 유효성 검사 로직을 함께 포함하는 '입력 단위 훅'**이라고 생각합니다.
> 
> 훅은 입력에 필요한 상태와 로직을 모두 포함하고 있어야 하며, UI 컴포넌트에서는 해당 훅을 연결하는 것만으로도 최소한의 코드로 동작하는 입력 필드를 만들 수 있어야한다는 생각이 들었습니다!
> 
> 이러한 고민 끝에, 이번에는 상태와 에러 메시지, 에러 상태를 모두 포함하고, setter와 함께 에러가 모두 통과했는지 여부와 입력값의 길이가 유효한지를 반환하는 하나의 커스텀 훅으로 통합하여 설계해 보았습니다.
> 이 방식에 대해서 동키콩은 어떻게 생각하시는지 궁금합니다!!

제 생각에 어떤 구조를 설계함에 있어서 도메인적 지식과 경험이 중요한 이유가 여기에서 나오는 것 같아요. 사실 응집도가 높은 구조를 설계하라는 말에서 응집도를 판단하는 기준에 있어서 변경사항에 대한 주관적인 판단이 들어가기 마련이고 변하는 부분과 변하지 않는 부분이 도메인마다 다르다고 생각해요. 때문에 호이초이가 2단계를 진행하고 상태로직과 에러로직은 동시에 움직이니까 같이 두는 것이 응집도가 높다고 판단한 부분 역시 경험에서 비롯된 합리적인 판단이라고 생각합니다

> 앞서 상태를 단일값과 배열값으로 변경하면서, 각각의 setter 함수를 setState 커스텀 함수 형태로 커스텀 훅 내부에서 정의하고, 이를 상위 컴포넌트로 전달하는 방식으로 리팩토링을 진행했습니다.
> 또한 에러 훅과 상태 훅을 결합하면서, 자연스럽게 유효성 검증과 에러 상태 및 메시지를 함께 처리하는 validateInput 함수도 해당 훅 내부에서 관리하도록 구성했습니다.
> 
> 기존에는 상태훅과 에러훅이 분리되어 있었기 때문에 setter와 validateInput을 inputField 컴포넌트에서 직접 조합해 onChange에 전달해야 했습니다. 하지만 현재는 동일한 훅 내부에서 이 두 로직을 통합해 내려주고 있습니다.
> 
> 이와 같은 흐름이라면, 처음부터 훅 내부에서 setter + 유효성 검증 로직을 조합해 하나의 onChange 핸들러로 내려주는 것이 더 깔끔한 설계일 것 같다는 생각이 듭니다. 혹시 이에 대해 동키콩은 어떻게 생각하실까요!?
> 
> 만약 이 방식이 더 적절하다면, 해당 핸들러 함수의 네이밍을 어떻게 정해야 할지도 고민이 됩니다. 예를 들어 단순한 handleChange보다 역할이 명확히 드러나는 이름으로 짓는 것이 좋을 것 같다고 생각 중입니다.

onChange함수를 훅내에서 처리하면 어떤지에 대한 질문으로 이해되는데요! onChange 핸들러를 훅 내부에 두는 것에는 장단점이 있다고 생각해요. 좋은 점은 컴포넌트 코드가 깔끔해져서 UI에 더 집중할 수 있다는 점이 있고, 관심사 분리가 확실히 될 것 같아요. 하지만 모든 onChange핸들러를 분리하는 것은 오히려 훅의 복잡도를 높이고 컴포넌트마다 조금씩 다른 onChange 동작을 구현해야 할 때 유연성이 떨어질 수 있을 것 같아요. 지금 정도라면은 개인적으로 분리해도 좋을 것 같다고 생각하는데요 그이유는 말씀하신대로 onChange로직에서 훅로직들을 통합하는 부분이 있기때문에 컴포넌트 내부에서 복잡도가 올라가는 부분이 있어서 관심사를 분리하는 것이 깔끔한 구조를 만들어 낼수 있다고 생각해서에요! 
이벤트에 직접적인 함수이기 때문에 handle~~Change도 괜찮다고 생각하구요. 또 update~~도 괜찮을 것 같아요!

JUDONGHYEOK reviewed (changes requested):
안녕하세요 호이초이! 이번 미션에서도 굉장히 많은 고민들을 해주셨네요 좋습니다! 덕분에 저도 많이 성장하는 느낌이에요!!

지금도 좋은 코드이고 훌륭한 구조라고 생각이 되어서 어프로브하고 싶은데요. 호이초이가 하고 싶은 말들이 남아 있는것 같아 코멘트 확인해주시고 다시 요청주시면 그때 어프로브하도록 하겠습니다!
2단계도 수고많으셨어요 호이초이!!

hoyyChoi committed:
refactor: useMemo 사용 formUIControllerData 값 저장

hoyyChoi committed:
refactor: FieldRenderer에서 prop에 대한 분기처리 작업 및 파일 분리

hoyyChoi committed:
fix: useStep 훅 step인자 제거

hoyyChoi commented:
안녕하세요! 동키콩🤡
말씀하신 수정사항들은 반영해서 개별메시지 커밋 달아두었습니다!!

> onChange함수를 훅내에서 처리하면 어떤지에 대한 질문으로 이해되는데요! onChange 핸들러를 훅 내부에 두는 것에는 장단점이 있다고 생각해요. 좋은 점은 컴포넌트 코드가 깔끔해져서 UI에 더 집중할 수 있다는 점이 있고, 관심사 분리가 확실히 될 것 같아요. 하지만 모든 onChange핸들러를 분리하는 것은 오히려 훅의 복잡도를 높이고 컴포넌트마다 조금씩 다른 onChange 동작을 구현해야 할 때 유연성이 떨어질 수 있을 것 같아요. 지금 정도라면은 개인적으로 분리해도 좋을 것 같다고 생각하는데요 그이유는 말씀하신대로 onChange로직에서 훅로직들을 통합하는 부분이 있기때문에 컴포넌트 내부에서 복잡도가 올라가는 부분이 있어서 관심사를 분리하는 것이 깔끔한 구조를 만들어 낼수 있다고 생각해서에요!
이벤트에 직접적인 함수이기 때문에 handleChange도 괜찮다고 생각하구요. 또 update도 괜찮을 것 같아요!

제가 잘못 설명드린 것 같습니다! 동키콩! 저는 훅내에서 `onChange` 함수를 관리하는 것은 전적으로 동의합니다!! 기존에는 `setter` 자체를 컴포넌트에 직접 내려주는 방식이었는데, 이 경우 하위 컴포넌트에서 상태를 임의로 조작할 수 있어 데이터 흐름이 불명확해지고 유지보수가 어려운 구조가 된다는 점을 직접 경험하면서 느꼈습니다. 특히, 배열 값을 단일 값으로 바꾸는 리팩토링을 진행할 때, 상위에서는 초기 값을 단일 값으로 변경했지만 하위 컴포넌트에서는 여전히 배열 방식으로 `setter`를 사용하는 로직이 남아 있어서, 예상치 못한 사이드 이펙트가 발생하기도 했습니다. 또한, 동키콩이 다른 크루에게 남긴 피드백 중 '**하위 컴포넌트에 직접 전달함으로써 캡슐화 원칙을 위반했을 때의 문제점**'에 대한 내용을 보며 많은 공감을 했고, 그 피드백을 바탕으로 저도 이 구조가 타당하다고 느꼈습니다!

다만, 제가 의문이 들었던 점은 입력값에 따라 상태를 바꿔주는 `onChange` 함수와 입력값을 검증하는 `validateInput` 함수가 현재 같은 훅 각각 정의되어 있는데, 이를 컴포넌트에 내려줄 때 이 둘을 각각 내려주는 것이 맞을지에 대한 부분이었습니다. 이 두 함수는 컴포넌트 딴에서 같은 인자들을 받고 작업을 진행하기 때문에 하나의 함수로 합쳐서 훅에서 `return` 시키고 이를 컴포넌트로 내려줘도 된다는 생각이 들었습니다. 이에 대한 동키콩의 의견이 궁금했습니다!!


## 추가 고민사항
이번에 리액트 모듈 작업을 병행하면서 훅 구조에 대해 고민이 많았고, 페이먼츠 미션 코드를 참고하면서 개선점을 고민해보았습니다.!
생각해보니 `isError`와 `errorMessage`는 결국 `uniqueNumber`라는 상태에 의존해 파생되는 값이기 때문에, 굳이 별도 상태로 분리하지 않아도 된다는 점이 떠올랐습니다.
불필요한 상태를 줄이는 것이 리액트의 기본 원칙이기도 해서, 아래와 같은 방식이 생각났는데, 동키콩은 어떻게 생각하실까요!?

### 현재 코드
```ts
  /** 상태 및 변수 관리 */
  const [uniqueNumber, setUniqueNumberState] = useState<UniqueNumberType>(["", "", "", ""]);
  const [isError, setIsError] = useState<ListErrorType>([false, false, false, false]);
  const [errorMessage, setErrorMessage] = useState("");


  const validateInput = (value: string, index: number) => {
    const { error, message } = validateNumberOnly(value);

    setIsError((prev) => {
      const updated = [...prev];
      updated[index] = error;
      return updated;
    });

    setErrorMessage(message);
  };
```


### 예상 코드
```ts
  const [uniqueNumber, setUniqueNumberState] = useState<UniqueNumberType>(["", "", "", ""]);

  /** 파생값 계산 */
  const errorInfoList = useMemo(() => uniqueNumber.map(validateNumberOnly), [uniqueNumber]);

  const isError = useMemo(() => errorInfoList.map(({ error }) => error), [errorInfoList]);

  const errorMessage = useMemo(() => {
    return errorInfoList.find(({ error }) => error)?.message || "";
  }, [errorInfoList]);
```

JUDONGHYEOK commented:
>제가 잘못 설명드린 것 같습니다! 동키콩! 저는 훅내에서 onChange 함수를 관리하는 것은 전적으로 동의합니다!! 기존에는 setter 자체를 컴포넌트에 직접 내려주는 방식이었는데, 이 경우 하위 컴포넌트에서 상태를 임의로 조작할 수 있어 데이터 흐름이 불명확해지고 유지보수가 어려운 구조가 된다는 점을 직접 경험하면서 느꼈습니다. 특히, 배열 값을 단일 값으로 바꾸는 리팩토링을 진행할 때, 상위에서는 초기 값을 단일 값으로 변경했지만 하위 컴포넌트에서는 여전히 배열 방식으로 setter를 사용하는 로직이 남아 있어서, 예상치 못한 사이드 이펙트가 발생하기도 했습니다. 또한, 동키콩이 다른 크루에게 남긴 피드백 중 '하위 컴포넌트에 직접 전달함으로써 캡슐화 원칙을 위반했을 때의 문제점'에 대한 내용을 보며 많은 공감을 했고, 그 피드백을 바탕으로 저도 이 구조가 타당하다고 느꼈습니다!

> 다만, 제가 의문이 들었던 점은 입력값에 따라 상태를 바꿔주는 onChange 함수와 입력값을 검증하는 validateInput 함수가 현재 같은 훅 각각 정의되어 있는데, 이를 컴포넌트에 내려줄 때 이 둘을 각각 내려주는 것이 맞을지에 대한 부분이었습니다. 이 두 함수는 컴포넌트 딴에서 같은 인자들을 받고 작업을 진행하기 때문에 하나의 함수로 합쳐서 훅에서 return 시키고 이를 컴포넌트로 내려줘도 된다는 생각이 들었습니다. 이에 대한 동키콩의 의견이 궁금했습니다!!

아하 제가 잘못 이해했었네요.😅 onChange함수와 validateInput을 하나로 합치는 하나의 훅을 만들면 어떻겠냐는 의견이신거죠? 오 저는 좋은 것 같습니다! 도메인 모듈들이 관심사에 따라 나뉘는 만큼 그 나눠진 모듈들을 다시 합쳐서 하나의 역할을 하는 모듈을 만드는 것은 자연스러운 흐름인 것 같아요! 이렇게 통합하면 컴포넌트 로직이 더 간단해지기도 하고 저는 좋은 생각인 것 같습니다!

> 이번에 리액트 모듈 작업을 병행하면서 훅 구조에 대해 고민이 많았고, 페이먼츠 미션 코드를 참고하면서 개선점을 고민해보았습니다.!
생각해보니 isError와 errorMessage는 결국 uniqueNumber라는 상태에 의존해 파생되는 값이기 때문에, 굳이 별도 상태로 분리하지 않아도 된다는 점이 떠올랐습니다.
불필요한 상태를 줄이는 것이 리액트의 기본 원칙이기도 해서, 아래와 같은 방식이 생각났는데, 동키콩은 어떻게 생각하실까요!?

이것도 역시 좋은 아이디어라고 생각해요! 파생상태라고 해서 무조건 useMemo를 사용하는 것이 좋다 이건 아니지만 현재 상황에서는 의존성배열이 그렇게 복잡한 경우가 아니고 복잡한 로직이나 외부 요인에 의존하는 경우도 아니기 때문에 충분히 useMemo로 활용하는 것도 좋은 것 같아요!!

JUDONGHYEOK reviewed (approved):
안녕하세요 호이초이!! 2단계 재 피드백도 빠르게 반영해주시고 답변을 남겨주셨네요! 추가로 질문주신 사항은 별도의 코멘트로 남겨두었으니 확인부탁드려요~!!

현재 모든 요구사항을 만족하면서 구조도 깔끔해진 것 같아 이만 어프로브 & 머지 하도록 하겠습니다!

2단계 미션도 호이초이와 즐거운 리뷰를 할 수 있어서 영광이였습니다~ 남은 미션들도 화이팅입니다!!
